File: apps/api/src/schemas/trip.py
from typing import Any, Literal

from pydantic import BaseModel, Field

from src.schemas.budget import BudgetSummary
from src.schemas.common import BookingLink, Confidence, SourceCitation, TravelPace
from src.schemas.event import EventOption
from src.schemas.flight import FlightOption
from src.schemas.hotel import HotelOption
from src.schemas.route import MapData


TripStatus = Literal["queued", "running", "completed", "needs_input", "failed"]
TripType = Literal[
    "full_trip",
    "weekend_plan",
    "local_event_search",
    "destination_inspiration",
    "itinerary_refinement",
    "booking_assistance",
]


class CurrentLocation(BaseModel):
    lat: float
    lng: float


class PlanTripRequest(BaseModel):
    query: str = Field(min_length=4, max_length=4000)
    origin: str | None = None
    destination: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    duration_days: int | None = Field(default=None, ge=1, le=45)
    traveler_count: int = Field(default=1, ge=1, le=12)
    budget: float | None = Field(default=None, ge=0)
    currency: str = Field(default="CAD", min_length=3, max_length=3)
    preferences: list[str] = []
    dietary_preferences: list[str] = []
    travel_style: TravelPace | None = None
    current_location: CurrentLocation | None = None


class RefineTripRequest(BaseModel):
    instruction: str = Field(min_length=3, max_length=2000)
    selected_option_id: str | None = None


class ProgressEvent(BaseModel):
    event: str
    label: str
    status: Literal["queued", "running", "completed", "failed"] = "completed"
    detail: str | None = None
    timestamp: str


class TripSummary(BaseModel):
    title: str
    origin: str | None = None
    destination: str | None = None
    duration_days: int | None = None
    traveler_count: int
    currency: str
    assumptions: list[str] = []


class ItineraryActivity(BaseModel):
    id: str
    start_time: str
    end_time: str
    title: str
    category: str
    description: str
    address: str
    lat: float
    lng: float
    estimated_cost: float
    currency: str
    travel_time_from_previous_minutes: int | None = None
    booking_required: bool = False
    booking_url: str | None = None
    source_provider: str
    confidence: Confidence = "medium"


class ItineraryDay(BaseModel):
    day_number: int
    date: str | None = None
    title: str
    theme: str
    activities: list[ItineraryActivity]


class ItineraryOption(BaseModel):
    id: str
    name: str
    best_for: str
    pace: TravelPace
    estimated_total: float
    currency: str
    confidence: Confidence
    highlights: list[str]
    flight: FlightOption | None = None
    hotel: HotelOption | None = None
    events: list[EventOption] = []
    days: list[ItineraryDay]
    budget_summary: BudgetSummary | None = None
    warnings: list[str] = []
    booking_links: list[BookingLink] = []


class FinalTripResponse(BaseModel):
    trip_id: str
    summary: TripSummary
    options: list[ItineraryOption]
    map_data: MapData
    booking_links: list[BookingLink]
    warnings: list[str]
    sources: list[SourceCitation]
    validation_report: dict[str, Any] | None = None


class PlanTripResponse(BaseModel):
    trip_id: str
    status: TripStatus
    follow_up_questions: list[str] = []
    result: FinalTripResponse | dict[str, Any] | None = None


class TripSessionResponse(BaseModel):
    trip_id: str
    status: TripStatus
    result: FinalTripResponse | dict[str, Any] | None = None
    progress_events: list[ProgressEvent] = []

File: apps/web/src/types/trip.ts
import type { EventOption } from "./event";
import type { FlightOption } from "./flight";
import type { HotelOption } from "./hotel";

export type TripStatus = "queued" | "running" | "completed" | "needs_input" | "failed";
export type TravelPace = "relaxed" | "balanced" | "packed";

export interface PlanTripRequest {
  query: string;
  origin?: string | null;
  destination?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  duration_days?: number | null;
  traveler_count: number;
  budget?: number | null;
  currency: string;
  preferences: string[];
  dietary_preferences: string[];
  travel_style?: TravelPace | null;
  current_location?: { lat: number; lng: number } | null;
}

export interface ProgressEvent {
  event: string;
  label: string;
  status: "queued" | "running" | "completed" | "failed";
  detail?: string | null;
  timestamp: string;
}

export interface ItineraryActivity {
  id: string;
  start_time: string;
  end_time: string;
  title: string;
  category: string;
  description: string;
  address: string;
  lat: number;
  lng: number;
  estimated_cost: number;
  currency: string;
  travel_time_from_previous_minutes?: number | null;
  booking_required: boolean;
  booking_url?: string | null;
  source_provider: string;
  confidence: "high" | "medium" | "low";
}

export interface ItineraryDay {
  day_number: number;
  date?: string | null;
  title: string;
  theme: string;
  activities: ItineraryActivity[];
}

export interface BudgetItem {
  category: string;
  amount: number;
  currency: string;
  confidence: "live" | "cached" | "estimated";
  notes: string;
}

export interface BudgetSummary {
  currency: string;
  items: BudgetItem[];
  subtotal: number;
  buffer: number;
  total_estimate: number;
  budget_status: "under_budget" | "near_budget" | "over_budget" | "unknown";
}

export interface BookingLink {
  label: string;
  url: string;
  provider: string;
  link_type: "flight" | "hotel" | "event" | "restaurant" | "activity";
  fetched_at: string;
}

export interface ItineraryOption {
  id: string;
  name: string;
  best_for: string;
  pace: TravelPace;
  estimated_total: number;
  currency: string;
  confidence: "high" | "medium" | "low";
  highlights: string[];
  flight?: FlightOption | null;
  hotel?: HotelOption | null;
  events: EventOption[];
  days: ItineraryDay[];
  budget_summary?: BudgetSummary | null;
  warnings: string[];
  booking_links: BookingLink[];
}

export interface MapMarker {
  id: string;
  label: string;
  type: "airport" | "hotel" | "attraction" | "restaurant" | "event";
  title: string;
  lat: number;
  lng: number;
  day?: number | null;
}

export interface RouteSegment {
  from_activity_id: string;
  to_activity_id: string;
  mode: "walking" | "driving" | "transit" | "mixed";
  distance_km: number;
  duration_minutes: number;
  polyline?: string | null;
  provider: string;
}

export interface MapData {
  center: { lat: number; lng: number };
  markers: MapMarker[];
  routes: { day: number; segments: RouteSegment[] }[];
}

export interface SourceCitation {
  provider: string;
  source_type: string;
  source_url?: string | null;
  fetched_at: string;
  confidence: string;
}

export interface FinalTripResponse {
  trip_id: string;
  summary: {
    title: string;
    origin?: string | null;
    destination?: string | null;
    duration_days?: number | null;
    traveler_count: number;
    currency: string;
    assumptions: string[];
  };
  options: ItineraryOption[];
  map_data: MapData;
  booking_links: BookingLink[];
  warnings: string[];
  sources: SourceCitation[];
  validation_report?: Record<string, unknown> | null;
}

export interface PlanTripResponse {
  trip_id: string;
  status: TripStatus;
  follow_up_questions: string[];
  result?: FinalTripResponse | null;
}
